test/api/core_test.dart

```diff
@@ -308,6 +321,41 @@ void main() {
       check(st.toString()).contains("distinctivelyNamedFromJson");
     }
   });
+
+  group('ApiConnection user-agent', () {
+    Future<void> checkUserAgent(String expectedUserAgent) async {
+      return FakeApiConnection.with_(account: eg.selfAccount,
+        useBinding: true,
+        (connection) async {
+          connection.prepare(json: {});
+          await connection.get(kExampleRouteName, (json) => json, 'example/route', null);
+          check(connection.lastRequest!).isA<http.Request>()
+            .headers.deepEquals({
+              ...authHeader(email: eg.selfAccount.email, apiKey: eg.selfAccount.apiKey),
+              ...{'User-Agent': expectedUserAgent},
+            });
+      });
+    }
+
+    final packageInfo = PackageInfo(version: '0.0.1', buildNumber: '1');
+
+    final testCases = [
+      ('ZulipFlutter/0.0.1+1 (Android 14)',             AndroidDeviceInfo(release: '14', sdkInt: 34),                        packageInfo),
+      ('ZulipFlutter/0.0.1+1 (iOS 17.4)',               IosDeviceInfo(systemVersion: '17.4'),                                packageInfo),
+      ('ZulipFlutter/0.0.1+1 (macOS 14.5.0)',           MacOsDeviceInfo(majorVersion: 14, minorVersion: 5, patchVersion: 0), packageInfo),
+      ('ZulipFlutter/0.0.1+1 (Windows)',                WindowsDeviceInfo(),                                                 packageInfo),
+      ('ZulipFlutter/0.0.1+1 (Linux; Fedora Linux 40)', LinuxDeviceInfo(name: 'Fedora Linux', versionId: '40'),              packageInfo),
+      ('ZulipFlutter/0.0.1+1 (Linux; Fedora Linux)',    LinuxDeviceInfo(name: 'Fedora Linux', versionId: null),              packageInfo),
+    ];
+
+    for (final (userAgent, deviceInfo, packageInfo) in testCases) {
+      test('matches $userAgent', () async {
+        testBinding.deviceInfoResult = deviceInfo;
+        testBinding.packageInfoResult = packageInfo;
+        await checkUserAgent(userAgent);
```

Comment:
Similar to my other comment about `testBinding.reset` and teardowns:
```suggestion
        testBinding.deviceInfoResult = deviceInfo;
        testBinding.packageInfoResult = packageInfo;
        addTearDown(testBinding.reset);
        await checkUserAgent(userAgent);
```

I think that suffices for taking the place of the global `tearDown` call.